#161342835 Create functionality to send JWT on signup

PriscillaSam opened the pull request:
#### What does this PR do?
- Create signup route.

#### Description of Task to be completed?
- Implement JWT for user authentication on successful signup
- Implement password hashing.
- Update user model to include `firstName` and `lastName`

#### How should this be manually tested?
- git clone `https://github.com/andela/haven-ah-backend.git`.
- Install `node` if not installed.
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.
To test on postman
- run `npm run start-dev`
- on postman navigate to `localhost:5000/api/users/signup`.

**Sample Request Body** 
```gherkhin
{
  "username": "string",
  "firstName": "string",
  "lastName": "string",
  "email": "string",
  "password": "string",
}
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#161342835

#### Screenshots (if appropriate)
<img width="1091" alt="screen shot 2018-10-26 at 10 54 52 am" src="https://user-images.githubusercontent.com/29674347/47559558-bdfbd400-d90d-11e8-88c6-edaafa27b7f5.png">

<img width="1080" alt="screen shot 2018-10-26 at 10 55 13 am" src="https://user-images.githubusercontent.com/29674347/47559525-aa506d80-d90d-11e8-9d6b-ded2f728d8b2.png">

PriscillaSam committed:
feat(signup): create signup functionality

- add migration to update user model with firstName and lastName fields
- add methods to create a new user
- add password hashing functionality
- add jwt generator function

[Finishes #161342835]

PriscillaSam committed:
feat(signup): create signup functionality

- add migration to update user model with firstName and lastName fields
- add methods to create a new user
- add password hashing functionality
- add jwt generator function

[Finishes #161342835]

wiztemple commented:
I love how the files are abstracted. its cool.

wiztemple commented:
The status code in the Postman don't correspond with the result.

ucheg6 commented:
Good job guys. However, there is a discrepancy between the status code on the snapshots of the postman and that of the returned objects.

TheoOkafor commented:
@wiztemple @ucheg6 the status code discrepancy will be fixed.

seyi-adeleke commented on package.json:
```diff
@@ -4,18 +4,22 @@
+    "start": "npm run build && node dist/app.js",
     "clean": "rm -rf dist/",
-    "test": "mocha --require babel-register server/tests/*.test.js",
+    "build": "npm run clean && mkdir dist/ && babel server/ -d dist/",
```
at what point are you going to run the pretest and posttests scripts?

PriscillaSam replied:
@seyi-adeleke the pretest script runs before the test script and postest script runs after the test script

seyi-adeleke replied:
so if i want to run the tests locally i'd have to run `npm run pretest &&  npm run test  && npm run posttest? `

PriscillaSam replied:
@seyi-adeleke not at all. By default calling `npm test` runs the pretest script first and the posttest after the test is run.

seyi-adeleke replied:
Nice 👍

seyi-adeleke commented on package.json:
```diff
@@ -25,23 +29,28 @@
     "mocha": "^5.2.0",
-    "nodemon": "^1.18.4"
+    "nodemon": "^1.18.4",
+    "pg": "^7.5.0",
```
pg and pg-hstore are repeated

seyi-adeleke commented on server/config/config.js:
```diff
@@ -18,12 +18,14 @@ const config = {
     use_env_variable: 'DATABASE_URL',
     PORT: process.env.PORT,
   },
+  secret: process.env.DB_PASSWORD,
```
nope. The JWT secret should be different for the three environments. Also it shouldnt be the same as the DB_PASSWORD

PriscillaSam replied:
Noted

seyi-adeleke commented on server/controllers/user.js:
```diff
@@ -0,0 +1,55 @@
+    } = request.body;
+
+    const user = await userRepo.getUserByEmail(email);
+    if (user) {
```
Can we create a formatter for our api?
It can be some sort of http utility. All you'll need to do is pass in the data and it should format it correctly. you'll probably need to pass in the response object as a parameter or something. Let me know if you have questions.

PriscillaSam replied:
@seyi-adeleke please this is not clear.

TheoOkafor replied:
Done

seyi-adeleke commented on server/repository/userRepository.js:
```diff
@@ -0,0 +1,39 @@
+  static async getUserByEmail(email) {
+    const user = await User.findOne({
+      where: { email },
+      attributes: {
```
shouldnt we exclude the password?

seyi-adeleke commented on server/routes/users.js:
```diff
@@ -0,0 +1,8 @@
+import { Router } from 'express';
+import User from '../controllers/user';
+
+const router = new Router();
```
You can probably extract `/api/v1` into a app.use statement

seyi-adeleke commented on server/controllers/user.js:
```diff
@@ -0,0 +1,55 @@
+   * @param {object} response Response Object
+   * @returns {object} User Object
+   */
+  static async signup(request, response) {
```
please integrate this with the validation middleware Victor wrote

PriscillaSam replied:
@seyi-adeleke the changes are currently on the develop branch. Once this is merged in, we would integrate the validation.

seyi-adeleke replied:
@PriscillaSam pull develop into your branch

PriscillaSam replied:
@seyi-adeleke noted.

seyi-adeleke commented on server/controllers/user.js:
```diff
@@ -0,0 +1,55 @@
+      lastName,
+      email,
+      username,
+      password: passwordHash,
```
Cant you just pass `hash(password)`  here?

seyi-adeleke commented on server/tests/controllerTests/user.test.js:
```diff
@@ -0,0 +1,36 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../../../app';
+import data from '../utilities/mockData';
```
We're not using a test db anymore?

PriscillaSam replied:
@seyi-adeleke we are. The mockData is just test data I'm sending to the db;

seyi-adeleke commented on server/utilities/jwtGenerator.js:
```diff
@@ -0,0 +1,12 @@
+import jwt from 'jsonwebtoken';
+import config from '../config/config';
+
+const webToken = (id) => {
```
jsdoc here. Also webToken isnt a very good name

TheoOkafor committed:
feat(signup): Create http response template

- Create http response utility functions
- Add input validation middleware to signup route
- Add middleware to remove extra whitespaces from strings

[Delivers #161342835]

seyi-adeleke commented on .coveralls.yml:
```diff
@@ -0,0 +1,2 @@
+ repo_token: process.env.COVERALLS_REPO_TOKEN
```
please remove the excess line here

seyi-adeleke commented on app.js:
```diff
@@ -16,6 +18,8 @@ app.use(logger('dev'));
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 
+app.use('/api', router);
```
Shouldnt this be `/api/vi` ?

PriscillaSam replied:
@seyi-adeleke Fixed.

seyi-adeleke commented on app.js:
```diff
@@ -16,6 +18,8 @@ app.use(logger('dev'));
 app.use(bodyParser.json());
```
What happened to bodyParser ?

PriscillaSam replied:
@seyi-adeleke I am not clear on your question.

seyi-adeleke commented on app.js:
```diff
@@ -16,6 +18,8 @@ app.use(logger('dev'));
+
 app.get('*', (request, response) => {
   const welcome = '<h1>Welcome to Authors Haven API Version 1.0</h1> ';
   response.status(200).send(welcome);
```
Let's use the http helper here #consistency

PriscillaSam replied:
@seyi-adeleke noted.

seyi-adeleke commented on package.json:
```diff
@@ -4,45 +4,57 @@
     "express": "^4.16.4",
     "jsonwebtoken": "^8.3.0",
     "morgan": "^1.9.1",
+    "nyc": "^13.1.0",
```
nyc should be a dev depedency

seyi-adeleke commented on server/config/config.js:
```diff
@@ -11,18 +11,22 @@ const config = {
     port: process.env.DB_PORT,
+    PORT: 5000,
     dialect: 'postgres',
+    secret: process.env.JWT_SECRET
```
Did you add this to travis?

PriscillaSam replied:
@seyi-adeleke  Yes I did

seyi-adeleke commented on server/config/config.js:
```diff
@@ -11,18 +11,22 @@ const config = {
   production: {
     use_env_variable: 'DATABASE_URL',
     PORT: process.env.PORT,
+    secret: process.env.JWT_SECRET
```
did you add this to Heroku?

PriscillaSam replied:
@seyi-adeleke We would add it.

seyi-adeleke commented on server/repository/userRepository.js:
```diff
@@ -0,0 +1,39 @@
+class UserRepository {
+  /**
+   * Function to create a user entity in the database
+
```
remove this space

seyi-adeleke reviewed (commented):
Good job. Address the comments and i'll merge this in

seyi-adeleke commented on server/utilities/cleanStrings.js:
```diff
@@ -0,0 +1,19 @@
+/**
```
Is this a middleware or a utility?

seyi-adeleke commented on server/utilities/cleanStrings.js:
```diff
@@ -0,0 +1,19 @@
+  Object.keys(request.body).forEach((key) => {
+    if (typeof request.body[key] === 'string') {
+      request.body[key] = request.body[key]
+        .replace(/  +/g, ' ')
```
Why cant you do `request.body[key].trim()` ?
